sequence_processing_pipeline/Pipeline.py

```diff
@@ -498,47 +495,49 @@ def _validate_mapping_file(mapping_file_path):
 
         try:
             df = pd.read_csv(mapping_file_path, delimiter='\t', dtype=str)
-
-            columns = [x.lower() for x in list(df.columns)]
-            if len(set(columns)) < len(columns):
-                raise PipelineError("Column names are case-insensitive. You "
-                                    "have one or more duplicate columns in "
-                                    "your mapping-file.")
-
-            # rename all columns to their lower-case versions.
-            # we will want to return this version to the user.
-            df.columns = df.columns.str.lower()
-
-            # if the two sets of headers are equal, verify there are no
-            # duplicate sample-names and then return the dataframe and no
-            # error message.
-            if len(exp - set(df.columns)) == 0:
-                # count the number of occurances of each sample-name.
-                dupes = df['sample_name'].value_counts()
-                # filter for duplicate sample-names
-                dupes = dupes.loc[lambda x: x > 1]
-                dupes = dupes.index.tolist()
-
-                if dupes:
-                    msg = ('duplicate sample-names detected: '
-                           '%s' % ', '.join(dupes))
-                else:
-                    return df
-            else:
-                if exp - set(df.columns) == exp:
-                    # since pandas.read_csv() will successfully open a sample-
-                    # sheet, if none of the expected columns exist then assume
-                    # that this file is not a mapping file.
-                    msg = "not a mapping-file"
-                else:
-                    msg = 'missing columns: %s' % ', '.join(
-                        exp - set(df.columns))
         except pd.errors.ParserError:
             # ignore parser errors as they obviously prove this is not a
-            # valid mapping file.
-            msg = 'could not parse file "%s"' % mapping_file_path
-
-        raise PipelineError(msg)
+            # valid mapping file. Return None w/error message
+            return None, 'Could not parse file "%s"' % mapping_file_path
+
+        columns = [x.lower() for x in list(df.columns)]
+        if len(set(columns)) < len(columns):
+            # return an invalid mapping file w/error message
+            return df, ("Column names are case-insensitive. You "
```

Comment:
In this code path, `is_mapping_file` would pass as it does not perform a check on an error message. It seems unexpected that `is_mapping_file` would evaluate `True` on an invalid mapping file. 

Is there any current use case in which the returned `df` is considered valid even if there are error messages?

Reply:
The issue here is that _validate_mapping_file() is called in two separate situations. In the first, we need to distinguish between a good mapping file, a bad mapping file, and something obviously not a mapping file (and more than likely a sample-sheet.) In the second, we just want to know whether it's a mapping-file or not (even if it's a bad one). This is to facilitate whether a file should be processed as amplicon or meta*omic and report any errors accordingly.

In this case, all is_mapping_file() is supposed to tell us is if we should process this as an amplicon pipeline or not. The  proper validation is handled downstream.